From 1497bd181de0a466a5c7d299bab47055aa2e9a05 Mon Sep 17 00:00:00 2001
From: Mihai Serban <mihai.serban@nxp.com>
Date: Wed, 9 Aug 2017 16:54:45 +0300
Subject: [PATCH 2362/5242] MLK-16177: soc: imx: Fix resources release in
 pm-domains detach_dev function

commit  c789d6670a6aeea26753981e64bf3b122a6544f2 from
https://source.codeaurora.org/external/imx/linux-imx.git

Because of a typo the resources allocated in imx8_attach_dev were not
correctly released by imx8_detach_dev.

Fixes: a0fb334819bb ("MLK16147-2 soc:imx Add support to save/restore clock rates")
Signed-off-by: Mihai Serban <mihai.serban@nxp.com>
Reviewed-by: Ranjani Vaidyanathan <Ranjani.Vaidyanathan@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/soc/imx/pm-domains.c |    6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/soc/imx/pm-domains.c b/drivers/soc/imx/pm-domains.c
index 5f94b47..90cd2df 100644
--- a/drivers/soc/imx/pm-domains.c
+++ b/drivers/soc/imx/pm-domains.c
@@ -182,15 +182,15 @@ static int imx8_attach_dev(struct generic_pm_domain *genpd, struct device *dev)
 static void imx8_detach_dev(struct generic_pm_domain *genpd, struct device *dev)
 {
 	struct imx8_pm_domain *pd;
-	struct imx8_pm_rsrc_clks *imx8_rsrc_clk;
+	struct imx8_pm_rsrc_clks *imx8_rsrc_clk, *tmp;
 
 	pd = container_of(genpd, struct imx8_pm_domain, pd);
 
 	/* Free all the clock entry nodes. */
-	if (!list_empty(&pd->clks))
+	if (list_empty(&pd->clks))
 		return;
 
-	list_for_each_entry(imx8_rsrc_clk, &pd->clks, node) {
+	list_for_each_entry_safe(imx8_rsrc_clk, tmp, &pd->clks, node) {
 		list_del(&imx8_rsrc_clk->node);
 		devm_kfree(dev, imx8_rsrc_clk);
 	}
-- 
1.7.9.5

